File: src/minima_harness/ai/events.py
"""Streaming event types emitted during assistant message generation.

A faithful port of PI's event taxonomy. Events are immutable dataclasses so they can
be safely fanned out to multiple subscribers. ``content_index`` associates each delta
or end event with its block (providers interleave deltas across text/thinking/tools).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from minima_harness.ai.types import AssistantMessage, StopReason, ToolCall

StreamEventReason = Literal["error", "aborted"]


@dataclass(frozen=True, slots=True)
class StartEvent:
    """Stream begins. ``partial`` is the initial assistant message skeleton."""

    type: Literal["start"] = "start"
    partial: AssistantMessage | None = None


@dataclass(frozen=True, slots=True)
class TextStartEvent:
    type: Literal["text_start"] = "text_start"
    content_index: int = 0


@dataclass(frozen=True, slots=True)
class TextDeltaEvent:
    type: Literal["text_delta"] = "text_delta"
    delta: str = ""
    content_index: int = 0


@dataclass(frozen=True, slots=True)
class TextEndEvent:
    type: Literal["text_end"] = "text_end"
    content: str = ""
    content_index: int = 0


@dataclass(frozen=True, slots=True)
class ThinkingStartEvent:
    type: Literal["thinking_start"] = "thinking_start"
    content_index: int = 0


@dataclass(frozen=True, slots=True)
class ThinkingDeltaEvent:
    type: Literal["thinking_delta"] = "thinking_delta"
    delta: str = ""
    content_index: int = 0


@dataclass(frozen=True, slots=True)
class ThinkingEndEvent:
    type: Literal["thinking_end"] = "thinking_end"
    content: str = ""
    content_index: int = 0


@dataclass(frozen=True, slots=True)
class ToolCallStartEvent:
    type: Literal["toolcall_start"] = "toolcall_start"
    content_index: int = 0


@dataclass(frozen=True, slots=True)
class ToolCallDeltaEvent:
    """Partial tool arguments (best-effort parse; fields may be missing)."""

    type: Literal["toolcall_delta"] = "toolcall_delta"
    delta: str = ""
    content_index: int = 0


@dataclass(frozen=True, slots=True)
class ToolCallEndEvent:
    type: Literal["toolcall_end"] = "toolcall_end"
    tool_call: ToolCall = None  # type: ignore[assignment]  # set by provider
    content_index: int = 0


@dataclass(frozen=True, slots=True)
class DoneEvent:
    type: Literal["done"] = "done"
    reason: StopReason = "stop"
    message: AssistantMessage = None  # type: ignore[assignment]  # set by provider


@dataclass(frozen=True, slots=True)
class ErrorEvent:
    """Emitted on provider error or abort. ``error`` carries partial content."""

    type: Literal["error"] = "error"
    reason: StreamEventReason = "error"
    error: AssistantMessage = None  # type: ignore[assignment]  # set by provider


Event = (
    StartEvent
    | TextStartEvent
    | TextDeltaEvent
    | TextEndEvent
    | ThinkingStartEvent
    | ThinkingDeltaEvent
    | ThinkingEndEvent
    | ToolCallStartEvent
    | ToolCallDeltaEvent
    | ToolCallEndEvent
    | DoneEvent
    | ErrorEvent
)

File: packages/tui/src/ai/events.ts
/**
 * Streaming events emitted during assistant message generation.
 *
 * Faithful port of the Python harness's ai/events.py (itself a port of PI's taxonomy).
 * `contentIndex` associates each delta/end event with its block — providers
 * interleave deltas across text/thinking/tools.
 */

import type { AssistantMessage, StopReason, ToolCall } from "./types.ts";

export type StreamEventReason = "error" | "aborted";

export interface StartEvent {
  type: "start";
  partial?: AssistantMessage;
}
export interface TextStartEvent {
  type: "text_start";
  contentIndex: number;
}
export interface TextDeltaEvent {
  type: "text_delta";
  delta: string;
  contentIndex: number;
}
export interface TextEndEvent {
  type: "text_end";
  content: string;
  contentIndex: number;
}
export interface ThinkingStartEvent {
  type: "thinking_start";
  contentIndex: number;
}
export interface ThinkingDeltaEvent {
  type: "thinking_delta";
  delta: string;
  contentIndex: number;
}
export interface ThinkingEndEvent {
  type: "thinking_end";
  content: string;
  contentIndex: number;
}
export interface ToolCallStartEvent {
  type: "toolcall_start";
  contentIndex: number;
}
export interface ToolCallDeltaEvent {
  type: "toolcall_delta";
  delta: string;
  contentIndex: number;
}
export interface ToolCallEndEvent {
  type: "toolcall_end";
  toolCall: ToolCall;
  contentIndex: number;
}
export interface DoneEvent {
  type: "done";
  reason: StopReason;
  message: AssistantMessage;
}
export interface ErrorEvent {
  type: "error";
  reason: StreamEventReason;
  error: AssistantMessage;
}

export type StreamEvent =
  | StartEvent
  | TextStartEvent
  | TextDeltaEvent
  | TextEndEvent
  | ThinkingStartEvent
  | ThinkingDeltaEvent
  | ThinkingEndEvent
  | ToolCallStartEvent
  | ToolCallDeltaEvent
  | ToolCallEndEvent
  | DoneEvent
  | ErrorEvent;

// Factories.
export const start = (partial?: AssistantMessage): StartEvent => ({ type: "start", partial });
export const textStart = (contentIndex: number): TextStartEvent => ({
  type: "text_start",
  contentIndex,
});
export const textDelta = (delta: string, contentIndex: number): TextDeltaEvent => ({
  type: "text_delta",
  delta,
  contentIndex,
});
export const textEnd = (content: string, contentIndex: number): TextEndEvent => ({
  type: "text_end",
  content,
  contentIndex,
});
export const thinkingStart = (contentIndex: number): ThinkingStartEvent => ({
  type: "thinking_start",
  contentIndex,
});
export const thinkingDelta = (delta: string, contentIndex: number): ThinkingDeltaEvent => ({
  type: "thinking_delta",
  delta,
  contentIndex,
});
export const thinkingEnd = (content: string, contentIndex: number): ThinkingEndEvent => ({
  type: "thinking_end",
  content,
  contentIndex,
});
export const toolCallStart = (contentIndex: number): ToolCallStartEvent => ({
  type: "toolcall_start",
  contentIndex,
});
export const toolCallDelta = (delta: string, contentIndex: number): ToolCallDeltaEvent => ({
  type: "toolcall_delta",
  delta,
  contentIndex,
});
export const toolCallEnd = (toolCall: ToolCall, contentIndex: number): ToolCallEndEvent => ({
  type: "toolcall_end",
  toolCall,
  contentIndex,
});
export const done = (reason: StopReason, message: AssistantMessage): DoneEvent => ({
  type: "done",
  reason,
  message,
});
export const error = (
  reason: StreamEventReason,
  assistantMessage: AssistantMessage,
): ErrorEvent => ({ type: "error", reason, error: assistantMessage });
